### PR TITLE
created mqtt authorization extension and created default oauth based authentication and carbon permission based authorization

### PR DESCRIPTION
**Pull request by ayyoob**

### CODE REVIEW

File: components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/CarbonPermissionBasedMQTTAuthorizer.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	@Override
+	public boolean isAuthorizedToConnect(MQTTAuthorizationSubject authorizationSubject) {
+		String permission = AuthorizationConfigurationManager.getInstance().getProperty(CONNECTION_PERMISSION_CONFIG);
+		if (permission != null && !permission.isEmpty()) {
+			String permissionAction = TreeNode.Permission.AUTHORIZE.toString().toLowerCase();
+			return isUserAuthorized(authorizationSubject, permission, permissionAction);
+		}
+		return true;
+	}
+
```
Comment (akalankapagoda) [inline components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/CarbonPermissionBasedMQTTAuthorizer.java]: Needs method comments.

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/internal/AuthenticationServiceComponent.java [hunk 2]
```diff
@@ -80,4 +96,23 @@ protected void setRealmService(RealmService realmService) {
...
+
+    protected void unsetServerConfiguration(ServerConfigurationService serverConfiguration) {
+    }
+
+    private int getPortOffset() {
+        ServerConfiguration carbonConfig = ServerConfiguration.getInstance();
+        String portOffset = System.getProperty("portOffset",
+                                               carbonConfig.getFirstProperty(CARBON_CONFIG_PORT_OFFSET));
+        try {
+            return ((portOffset != null) ? Integer.parseInt(portOffset.trim()) : CARBON_DEFAULT_PORT_OFFSET);
```
Comment (akalankapagoda) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/internal/AuthenticationServiceComponent.java]: Would it be more clear if we break this into an if else statement.

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	 */
+	@Override
+	public void passivateObject(Object o) throws Exception {
+		if (o instanceof OAuth2TokenValidationServiceStub) {
+			OAuth2TokenValidationServiceStub stub = (OAuth2TokenValidationServiceStub) o;
+			stub._getServiceClient().cleanupTransport();
+		}
+	}
+
+	private OAuth2TokenValidationServiceStub generateStub() throws OAuthTokenValidationException {
```
Comment (hemikak) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java]: Method comments needed.

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/OAuth2BasedMQTTAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+				log.warn("Error occurred while returning the object back to the oauth token validation service " +
+								 "stub pool.", e);
+			}
+		}
+		AuthenticationInfo authenticationInfo = new AuthenticationInfo();
+		authenticationInfo.setAuthenticated(false);
+		return authenticationInfo;
+	}
+
+	private AuthenticationInfo getAuthenticationInfo(String token,
```
Comment (hemikak) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/OAuth2BasedMQTTAuthenticator.java]: This line needs to be formatted. Method doc comments needed.

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/OAuth2BasedMQTTAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+public class OAuth2BasedMQTTAuthenticator implements IAuthenticator {
+	private static final Logger log = Logger.getLogger(OAuth2BasedMQTTAuthenticator.class);
+	private static final String TOKEN_TYPE = "bearer";
+	private static final String SCOPE_IDENTIFIER = "scope";
+	private static final String TOKEN_EXPIRY_TIME_IDENTIFIER = "expiry_time";
+	private static String cookie;
+	private GenericObjectPool stubs;
+
+	/**
+	 * initialize the OAUTH2ValidationStubFactory  to communicate with the OAuth2TokenValidationService
```
Comment (hemikak) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/OAuth2BasedMQTTAuthenticator.java]: *Initial  .

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+		return this.generateStub();
+	}
+
+	/**
+	 * This is used to clean up the Oauth validation stub and releases to the object pool.
+	 * @param o
+	 * @throws Exception
+	 */
+	@Override
+	public void passivateObject(Object o) throws Exception {
```
Comment (hemikak) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java]: Need to describe the params in the doc.

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+					options.setProperty(HTTPConstants.AUTHENTICATE, auth);
+					options.setProperty(HTTPConstants.REUSE_HTTP_CLIENT, Constants.VALUE_TRUE);
+					client.setOptions(options);
+					if (hostURL.getProtocol().equals("https")) {
+						// set up ssl factory since axis2 https transport is used.
+						EasySSLProtocolSocketFactory sslProtocolSocketFactory =
+								createProtocolSocketFactory();
+						Protocol authhttps = new Protocol(hostURL.getProtocol(),
+														  (ProtocolSocketFactory) sslProtocolSocketFactory,
+														  hostURL.getPort());
```
Comment (hemikak) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java]: This line needs to be formatted.

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+		}
+
+		return stub;
+	}
+
+	/**
+	 * @return an EasySSLProtocolSocketFactory for SSL communication. This is required because of using
+	 * CommonHTTPTransport(axis2 transport) in axis2
+	 */
+	private EasySSLProtocolSocketFactory createProtocolSocketFactory() throws OAuthTokenValidationException{
```
Comment (hemikak) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java]: Need a comment as to what this method performs

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+			String errorMsg = "Failed to set the key material in easy ssl factory.";
+			throw new OAuthTokenValidationException(errorMsg, e);
+		}
+
+	}
+
+	/**
+	 * @return an instance of HttpClient that is configured with MultiThreadedHttpConnectionManager
+	 */
+	private HttpClient createHttpClient() {
```
Comment (hemikak) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java]: Need a doc comment to what this method performs.

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/exception/OAuthTokenValidationException.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.andes.authentication.andes.oauth.exception;
+
+public class OAuthTokenValidationException extends Exception {
```
Comment (hemikak) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/exception/OAuthTokenValidationException.java]: Need class comment.

File: components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/CarbonPermissionBasedMQTTAuthorizer.java (hunk 1 above)
Comment (ayyoob) [inline components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/CarbonPermissionBasedMQTTAuthorizer.java]: Added the comments

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/internal/AuthenticationServiceComponent.java (hunk 2 above)
Comment (ayyoob) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/internal/AuthenticationServiceComponent.java]: Did break it to if else statement

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java (hunk 3 above)
Comment (ayyoob) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java]: added the comments

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/OAuth2BasedMQTTAuthenticator.java (hunk 4 above)
Comment (ayyoob) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/OAuth2BasedMQTTAuthenticator.java]: Added the comments and formatted

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/OAuth2BasedMQTTAuthenticator.java (hunk 5 above)
Comment (ayyoob) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/OAuth2BasedMQTTAuthenticator.java]: Changed it.

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java (hunk 6 above)
Comment (ayyoob) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java]: described it

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java (hunk 7 above)
Comment (ayyoob) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java]: fixed the formatting issue.

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java (hunk 8 above)
Comment (ayyoob) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java]: Added a comment

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java (hunk 9 above)
Comment (ayyoob) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/OAuthTokenValidaterStubFactory.java]: Add a comment

File: components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/exception/OAuthTokenValidationException.java (hunk 10 above)
Comment (ayyoob) [inline components/andes/org.wso2.carbon.andes.authentication/src/main/java/org/wso2/carbon/andes/authentication/andes/oauth/exception/OAuthTokenValidationException.java]: Added a class level comment

**Pull request merged**